File: archive-pre-google/remove-duplicates-from-sorted-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
'''
oh
1 2 3
    h
'''
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        orighead=head
        while head:
            while head.next and head.next.val==head.val:
                head.next=head.next.next
            head=head.next
        return orighead

File: lc/remove-duplicates-from-sorted-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// d 1 -1 1 2 3 n
//   p-------
//           c
// so p moves only when they are not the same
// c always moves
// d 1 1 2
// d 1   2
//   --- p
//         c
func deleteDuplicates(head *ListNode) *ListNode {
    d := &ListNode {Val: -101, Next: head}
    p := d
    for c := head; c != nil; c = c.Next {
        if p.Val == c.Val { 
            p.Next = c.Next
        } else {
            p = c
        }
    }
    return d.Next
}